### ISSUE TITLE
mc 1.12.2 - Placement preview missing ...

### ISSUE THREAD

**Issue by P3rf3ctXZer0** [Direwolf20-MC/BuildingGadgets#329]
**IMPORTANT:** Before continuing, please ensure that Building Gadgets (and any other mods involved) is updated to the latest available version
----------------------------------------------------------------------------
After deleting this template, please provide the following information:
* Forge version Forge 14.23.5.2838
* Building Gadgets version BuildingGadgets-2.7.0

Placement preview missing?
[latest.log](https://github.com/Direwolf20-MC/BuildingGadgets/files/3422471/latest.log)

**Comment by P3rf3ctXZer0** [comment]
Working on getting video ... I have to fix OBS

**Comment by P3rf3ctXZer0** [comment]
Here you go https://youtu.be/zpDusHF3BZE

**Comment by MajorTuvok** [comment]
That's odd, because it is present in dev. Are you sure you did that correctly, because I cannot see what keys you are pressing.

**Comment by MichaelHillcox** [comment]
It builds so it's kinda clear that he has a block selected. It would look to me as this could be cause by another mod. Not sure though. There is a mod present modifying the render of the bounding box of elements although this logically shouldn't effect our render.

**Comment by P3rf3ctXZer0** [comment]
See if it is an issue between effortless building and builders gadgets.

**Comment by MichaelHillcox** [comment]
Does this issue happen when effortless building is in normal mode?

**Comment by P3rf3ctXZer0** [comment]
Yeah

**Comment by MichaelHillcox** [comment]
Shouldn't be an issue. What are the numbers in the centre of your screen?

**Comment by P3rf3ctXZer0** [comment]
Error nothing in the center of my screen do you mean lower right?

**Comment by P3rf3ctXZer0** [comment]
Oh distance from blocks XD

**Comment by MichaelHillcox** [comment]
In the video. There are numbers when looking at your crops?

**Comment by MichaelHillcox** [comment]
Oh. Then I literally have no clue what could be interfering with the render. It's working fine on my end. What pack are you running?

**Comment by P3rf3ctXZer0** [comment]
I am a builder ... I only use it when counting blocks needed for structures :D

**Comment by P3rf3ctXZer0** [comment]
Do you have an 8gb pc ? if so you can jump on my server and see things first hand.

**Comment by P3rf3ctXZer0** [comment]
6Gb for pack and 2 gb for windows os.

**Comment by MichaelHillcox** [comment]
I'd prefer to test locally

**Comment by P3rf3ctXZer0** [comment]
Um ... hmm ... I can print my mod directory.

**Comment by P3rf3ctXZer0** [comment]
[file-tree-dump.txt](https://github.com/Direwolf20-MC/BuildingGadgets/files/3423157/file-tree-dump.txt)

**Comment by MichaelHillcox** [comment]
Just wanted to know what mod pack it is. Is it something you've made yourself?

**Comment by MichaelHillcox** [comment]
Just wanted to know what mod pack it is. Is it something you've made yourself?

**Comment by P3rf3ctXZer0** [comment]
Its a server I host for friends and people interested.

**Comment by MajorTuvok** [comment]
Can you try adding/removing mods until it no longer hapoens for you? Also on our end we should propably additionally test with the Forge he is using...(because Forge hasn't released another recommended build, we are still stuck with this pretty old one)

**Comment by P3rf3ctXZer0** [comment]
Maybe ... it is a low priority though because of the 1.14.4 release and focus. I am only staying on 1.12.2 until pixelmon gets pushed to 1.14.4

**Comment by MajorTuvok** [comment]
As I said we'll be releasing a 1.14 4 alpha in a few days too...

**Comment by P3rf3ctXZer0** [comment]
Fixed in latest version 
BuildingGadgets-2.7.1.jar

Closing issue due to not being an issue

**Comment by azaghal** [comment]
Hm... Don't like resurrecting closed issues, but... I actually am having the same issue on Enigmatica 2: Expert modpack. The mods have not been modified in the pack, _except_ for the addition of Lag Goggles (`LagGoggles-FORGE-1.12.2-4.3-HOTFIX.jar`) - which normally has not caused me issues in the past (but, you never know, could be some version bump issues).

This has started happening after upgrade from modpack version 1.64d (which had Building Gadgets 2.6.8) to 1.66c (which has building gadgets 2.7.1).

 I just tested on a new world in creative mode as well to see if it persists.

Looking at mod differences between the old and upgraded _modpack_ version, they all seem to be version bumps, except the following:

- overloadedarmorbar-1.0.4g.jar
- toughnessbar-2.3.jar

I'll try to play around a tiny bit with the listed new mods (including Lag Goggles) - I can see the OP had both the overloadedarmorbar and Lag Goggles, incidentally :)

**Comment by P3rf3ctXZer0** [comment]
Hey so apparently not fixed then

**Comment by azaghal** [comment]
Ok, so... Disabling the overloadedarmorbar-1.0.4g.jar, toughnessbar-2.3.jar and Lag Goggles did not help. Then I went ahead to see what happens if I downgrade Building Gadgets to 2.6.8, and... That one did help. I'll try setting-up a clean profile with just Building Gadgets for start. And, given the hint from OP, might be useful for me to look into those couple of mods that deal with rendering after that too :)

**Comment by azaghal** [comment]
I'll try adding mods to empty profile in alphabetical groups to see if I can reproduce problem at some point. I'll let you know if I managed to find something.

**Comment by azaghal** [comment]
Now, I might be a bit biased because I traditionally play as a dwarf in RPGs, but I think it's the elves' fault... They must be trying to break through from Alfheim...

Witty comments aside, I was able to reproduce this with the following:

- Minecraft 1.12.2
- Forge 14.23.5.2836
- Baubles-1.12-1.5.2.jar
- Botania r1.10-361.jar
- BuildingGadgets-2.7.1.jar

**Comment by MajorTuvok** [comment]
I'm sorry, but I'm very confused... We can look into it however...

**Comment by azaghal** [comment]
If it's about the elf thing - it's just a bit of bad humor on my side - I'll refrain myself in the future :)

I should have mentioned that disabling the Botania mod (leaving Baubles turned on) seems to "solve" the rendering issue.

If anyone else could reproduce this with listed minimal set of mods - including verifying that disabling solves the problem (maybe @P3rf3ctXZer0?), that could be a good thing (just so it's not one of the "well, on my machine..." things :).

**Comment by MajorTuvok** [comment]
I'm confused as to why other mods disable our render... The only thing I can think of is that they cancel the render event or we crash when using their blocks. But both doesn't seem to be the case.

**Comment by azaghal** [comment]
Well, sounds a bit weird, it does. Now, what's interesting is that downgrading to 2.6.8 of Building Gadgets "solves" the issue too.

I saw the changelog for 2.7.0 mentioning a lot of changes in the code, though, so has rendering handling seen major updates as well?

**Comment by MajorTuvok** [comment]
I'm pretty sure that there wern't many changes to the render - rather to what is rendered which I don't see why that should be interfering with other mods. We'll need to do quite a bit of debugging I guess...

**Comment by MichaelHillcox** [comment]
There are no functional chnages to the rendering code. Its was all code cleaning :)

**Comment by MichaelHillcox** [comment]
I can't reproduce this in my dev environment with the mods outlined. can you send me your world save?

**Comment by P3rf3ctXZer0** [comment]
Strange cause since the update - it has been working great and you saw first hand @MichaelHillcox that mine now works fine.

**Comment by MichaelHillcox** [comment]
What, 2.7.1 fixed it?

**Comment by P3rf3ctXZer0** [comment]
Yes.

**Comment by P3rf3ctXZer0** [comment]
This is the version I am using that works BuildingGadgets-2.7.1.jar

**Comment by P3rf3ctXZer0** [comment]
These are all the mods I have.
[file-tree-dump.txt](https://github.com/Direwolf20-MC/BuildingGadgets/files/3464214/file-tree-dump.txt)
Please ignore name adjustments.

**Comment by MichaelHillcox** [comment]
This is all pointless information if it works for you

**Comment by P3rf3ctXZer0** [comment]
Trying to help the guy who had me re-open this issue. @azaghal

**Comment by azaghal** [comment]
Well, I'm a GNU/Linux user, so that command in particular won't work for me - here is a find combined with a bit of checksums:

```
find . -type f -exec sha256sum '{}' \;
765a2f617b54e7cdcf77ef2c765b6fa59f08d7bc51259643aeefdecb6cd8bbb2  ./BuildingGadgets-2.6.8.jar.disabled
42725a364da12a72f7f3139ec0fb758dcd8f43026397ca03fb6dd81f52bd9b63  ./Botania r1.10-361.jar
50f5af55f6e2b33dc6a4dcc1085f31af031b862cdbc315b47eeac20b5ff710bf  ./BuildingGadgets-2.7.1.jar
b32010b2f2778aa1188585e7ead91ad46d4cb2c715f9c778a61848ba7fe51f8d  ./Baubles-1.12-1.5.2.jar
```

I can guarantee the directory contents and the steps taken to reproduce :)

I'm using MultiMC for running the modded instances, not sure if that would help in any way. And the instance/world I have tested on was brand new - so no stale data should be present. Take note that the disabled 2.6.8 version was added subsequently so I could go back and forth between versions for testing (and was not present in the first run).

I'll retest tomorrow once again, just to be on the safe side, but I'm pretty confident in reproduction steps. Should I try using the most up-to-date version of Forge as well? I've tried using what's shipping with my Enigmatica 2: Expert instance.

Looking at @P3rf3ctXZer0 latest list of mods - looks like more stuff got updated along the way, so could be more up-to-date forge or what-not as well. _Maybe_ :)

**Comment by P3rf3ctXZer0** [comment]
I am using latest 1.12.2 because the only draw back is some devs won't fix broken code and even though the forge team are not officially saying 1.12.2 is done just look how much effort is being poured into 1.1.4.4

**Comment by MichaelHillcox** [comment]
Only thing to note is that you're on a newer forge version than my dev environment. Is there a specific reason you're not using the recommended build?

**Comment by P3rf3ctXZer0** [comment]
That would be Quark and JEI's fault and Lexmanos who basically greenlite but didn't care to actually make it the recommended version. Its because he and his entire team are hyper focused on 1.14.4

**Comment by P3rf3ctXZer0** [comment]
there is beyond a shadow of a doubt nothing wrong in 838

**Comment by MichaelHillcox** [comment]
That was directed at @azaghal 
@P3rf3ctXZer0 it's unlikely that those two mods require the non recommended version.

**Comment by P3rf3ctXZer0** [comment]
Grab a copy and run it yourself. https://www.curseforge.com/minecraft/mc-mods/quark/files/2753054

**Comment by P3rf3ctXZer0** [comment]
838 added specific fixes which Quark and JEI immediately jumped on especially since they requested the fixes in the first place.

**Comment by azaghal** [comment]
@MichaelHillcox The Enigmatica 2: Expert modpack depended on it - I _think_ it was Actually Additions contained within that complained during my initial testing (when I started adding mods used in Enigmatica to new instance in smaller batches).

What version of Forge would you like me to try with specifically (MultiMC marks `Forge 14.23.5.2768` as recommended on my side)?

I'm happy to do testing as much as it's possible to pin this down if possible - narrowing this down on my side to Botania + Building Gadgets immensely reduces startup time for testing :)

**Comment by MajorTuvok** [comment]
You already said, that you are running a Linux System? What JVM and JVM version are you using? (I'm asking so that I can do some googling on whether there might be JIT issues on the JVM)

**Comment by azaghal** [comment]
From `java -version`:

```
openjdk version "1.8.0_222"
OpenJDK Runtime Environment (build 1.8.0_222-8u222-b10-1~deb9u1-b10)
OpenJDK 64-Bit Server VM (build 25.222-b10, mixed mode)
```

I have also successfully reproduce the issue from scratch again (just to make sure I wasn't doing something funky with versioning :)

1. Created new profile in MultiMC taking Minecraft 1.12.2 as version.
2. Upped the maximum memory allocation to 4096MB for the instance (just in case). All other JVM parameters kept as-is (what the MultiMC uses by default).
3. LWJGL 2 is at version 2.9.4-nightly-20150209 (not sure if this is relevant, but since it's listed in profile might as well mention it).
4. Installed `Forge 14.23.5.2768` (recommended version by MultiMC).
5. Added mods (listing checksums):

```
b32010b2f2778aa1188585e7ead91ad46d4cb2c715f9c778a61848ba7fe51f8d  Baubles-1.12-1.5.2.jar
42725a364da12a72f7f3139ec0fb758dcd8f43026397ca03fb6dd81f52bd9b63  Botania r1.10-361.jar
50f5af55f6e2b33dc6a4dcc1085f31af031b862cdbc315b47eeac20b5ff710bf  BuildingGadgets-2.7.1.jar
```
6. Started instance, created new world, creative, superflat.
7. Gave myself building gadget and stone slabs.
8. Placed stone slabs in the world, picked them with sneak-right click and placed some using building gadget. They do place in the world, but no preview.

**Comment by MajorTuvok** [comment]
It's unlikely that this is caused by the JVM then - I can find nothing and the only issues I knew of were fixed somewhere around 50... @MichaelHillcox If you can't reproduce, I'll try standard MC launcher later on and see what happens

**Comment by MichaelHillcox** [comment]
I'll do the same soon

**Comment by azaghal** [comment]
@P3rf3ctXZer0 Would it be simple/possible for you to try out reproducing issue using the minimal set of mods and versions I listed (no need for MultiMC, just use whatever you're normally using, I'd say)? Asking since you had the same problem before, and you are in addition on different platform compared to me.

**Comment by MichaelHillcox** [comment]
We're trying to reproduce it in our dev environments so we have the ability to debug the code in real time to find the issue. At the moment, with the mod set stated, we're not able to reproduce your issue. We'll try again outside of our dev environments but it does make it a lot more annoying to try and fix :P

**Comment by azaghal** [comment]
No worries - I really appreciate all the effort, so thanks for still trying :)

**Comment by P3rf3ctXZer0** [comment]
@azaghal can I please have a complete copy of you Multimc instance is it was when you started this report? Please change nothing?

**Comment by Moleculor** [comment]
So, I don't know how much I can assist, if at all, as the very start of this issue requests that I ensure I'm on the latest forge version of 2838, but Enigmatica 2: Expert has Forge 2836, but I'm having this exact issue with the Building Gadget (but not the Exchanging Gadget) not showing previews. 

This is my very first time using this mod, so I thought I was doing something wrong. 

I'd be happy to zip up my entire directory and send it somewhere, if it'll help. Minus the backups, it's ~432MB, so I'm not sure HOW I'd send it, but I'd be willing to.

**Comment by MichaelHillcox** [comment]
Upload it to mega, gdrive, Dropbox and put a link here or you can email me it at me@mikey.pro

**Comment by MajorTuvok** [comment]
I managed to reproduce it!!!!

**Comment by MajorTuvok** [comment]
It only happens in superflat worlds for me and I needed to downgrade my dev Forge and re upgrade it, in order to get this issue, but I've got it in my dev env now

**Comment by MajorTuvok** [comment]
Moment sorry, I'm just dump... I attempted to use build to me in the bottom XD

**Comment by Penetrosity** [comment]
Enigmatica 2: Expert after the update (Which removed dank/null, and probalby did more) from E2:E 1.66c to 1.67.

The Forge version didn't change `Using Forge-1.12.2-14.23.5.2836`
The changelog doesn't state it changes the Building Gadgets version. Current version I'm using `BuildingGadgets-2.7.1.jar`
The only mod I've added/changed out of the standard E2:E download is `LagGoggles-FAT-1.12.2-4.6.jar`

Here is my current folder. The only thing I removed is ReAuth.cfg because it had my personal email listed within. (I definitely didn't it have it save my password in a text file haHAA)

https://drive.google.com/file/d/1d4Q_sabP5Bsp4hM9b2m1b7D8shgw5DVv/view?usp=drivesdk

I can also take video or whatever else is needed, just let me know, thanks!

**Comment by loordgek** [comment]
https://paste.ee/p/PyjmE
i found it, it is Botania r1.10-361.jar

**Comment by MichaelHillcox** [comment]
I couldn't replicate it with my dev environment

**Comment by loordgek** [comment]
with out botania
![with out botania](https://user-images.githubusercontent.com/4789835/62649984-20323c80-b956-11e9-9b75-3f6997250821.png)

with botania
![with botania](https://user-images.githubusercontent.com/4789835/62650027-36d89380-b956-11e9-8395-31a5ba94ffee.png)

**Comment by MichaelHillcox** [comment]
I love your background

**Comment by Moleculor** [comment]
Based on @loordgek's comment, I tried deleting Botania r1.10-361.jar from my Enigmatica 2: Expert instance. 

Previews did not show up. 

I've emailed @MichaelHillcox a copy of my MultiMC instance.

**Comment by MajorTuvok** [comment]
I also tried adding botnia in dev env and my build views showed up as expected...
I guess we @MichaelHillcox need to try adding all configs and mods to dev which are also in e2e and see what happens?

**Comment by loordgek** [comment]
downgrading form 2.7.1 > 2.6.8 fixed the problem

**Comment by MajorTuvok** [comment]
It's definitly the rewrite... But where and why...?!?

**Comment by P3rf3ctXZer0** [comment]
Weird mine was the opposite it was 2.7.1 < 2.6.8

**Comment by P3rf3ctXZer0** [comment]
Maybe issue is a bit further back?

**Comment by loordgek** [comment]
it renders sometimes really weird 
![Untitled1](https://user-images.githubusercontent.com/4789835/62655237-d6e7ea00-b961-11e9-91c8-0a409a62e309.png)

**Comment by MajorTuvok** [comment]
Moment SocketTimeout? @loordgek could you send me this SocketTimeout log?

**Comment by MajorTuvok** [comment]
Even though that might come from someone else it's worth trying...
@MichaelHillcox cause I won't have any time tomorrow, could you compile a jar which has some logs in all render steps to see whether they get called (and maybe even print interesting data) and then @loordgek @P3rf3ctXZer0 or someone else could run with that and send us the resulting log?

**Comment by P3rf3ctXZer0** [comment]
Yeah I can do that - no problem :D @MajorTuvok

**Comment by azaghal** [comment]
@P3rf3ctXZer0 Sorry for getting a bit late to you, but I could prepare a fresh instance with just Botania and Building Gadgets if you'd like (one where I'd also test to make sure the issue persists)? I would say it's a pretty bare bones reproduction environment that has triggered the bug for me every single time.

Has anyone actually managed to reproduce the issue with just Baubles, Botania, and Building Gadgets except me?

Btw, @loordgek 's comment about weird rendering is interesting - the preview started working for me at some point, and it was kinda same "weird rendering" thing as well. After the client restart, though, it stopped working again (and I could not figure out what made it suddenly work either).

**Comment by loordgek** [comment]
@MajorTuvok cant find that log anymore :(

the exchanging gadget and the copy/paste tool preview render seems to to work just fine

**Comment by loordgek** [comment]
some blocks it is trying to render coblestone/brick stair it is rendering a translucent white shape of the stair wood is not rendering
for coblestone and cactus it renders a translucent white block 
some leafs renders as a translucent green block 
some other it renders something

**Comment by MajorTuvok** [comment]
Sounds like the BlockProvider returns air - _-

**Comment by loordgek** [comment]
is is rendering it consistent  i have restarted 10 times and it stays the same

**Comment by loordgek** [comment]
why confused ?

**Comment by MajorTuvok** [comment]
If anyone of you could find his or her log...if you'd be using a dedicated Server I'd ask whether your client recieved a message `recieved config from server`? And if not whether the durability configs differ...

**Comment by MajorTuvok** [comment]
I'm confused because this issue doesn't seem to happen every start - why the hack does it happen only sometimes?!?

**Comment by MajorTuvok** [comment]
Does the aforementioned lib still fail to get it it's contributors list?

**Comment by loordgek** [comment]
no

**Comment by Moleculor** [comment]
Well, when @MichaelHillcox gets my instance folder, y'all probably can replicate it. Unless someone else wants me to send it to them? I'm not fully comfortable just throwing the file link on Github, but I can't seem to find any private messaging on here for the people marked as collaborators and authors.

**Comment by loordgek** [comment]
https://michaelhillcox.com/contact

**Comment by Moleculor** [comment]
I already emailed him.

**Comment by MajorTuvok** [comment]
Hmm, I need to check whether they close their sockets after crash.

@Moleculor one can find us on Dire's discord server too. Also I don't have the time to test that today even if you send it to me...

**Comment by loordgek** [comment]
> Hmm, I need to check whether they close their sockets after crash.

i dont think that is a issue i am only using atm botania baubles and this forge-1.12.2-14.23.5.2768

i am of to work now

**Comment by MajorTuvok** [comment]
Hmm, ok thanks

**Comment by MajorTuvok** [comment]
Dire just remembered that he had a very similiar if not the same issue with botania already and applied a fix for it which is already in the [Exchanger-Render](https://github.com/Direwolf20-MC/BuildingGadgets/blob/712f602e91f1e15694aeb6b4025e3cee8517d45c/src/main/java/com/direwolf20/buildinggadgets/common/tools/ToolRenders.java#L176). We talked a bit about it and because no one knew why it was there (looked like useless code) and there was no comment, we removed it (apperently only for the builder)..

According to Dire's old discussion with Vazkii this is some botania issue though - _-. We'll apply the fix nonetheless of course.

**Comment by Moleculor** [comment]
Ahahahaha.

Okay, THIS time, add comments.

**Comment by MajorTuvok** [comment]
Released in Version 2.7.2 a few min ago
